### PR TITLE
Rewrite parse generate

### PR DESCRIPTION
**Pull request by jdidion**
This PR is a re-write of the code used to parse WDL source and generate native apps/workflows. Most importantly, it uses wdlTools parser to parse WDL and wdlTools code generator to generate WDL from typed AST. All the source and tests compile, but I need to:

* Merge the latest changes from TNG_II
* Get the tests to pass

### CODE REVIEW

File: src/main/scala/dxWDL/base/Extras.scala [hunk 1]
```diff
@@ -689,18 +679,16 @@ object Extras {
...
     }
-
   }
 
   def parseCustomReorgAttrs(jsv: JsValue, verbose: Verbose): Option[ReorgAttrs] = {
-    if (jsv == JsNull)
+    if (jsv == JsNull) {
       return None
-
+    }
```
[the remark this answers is not in the document]
Comment (jdidion) [reply src/main/scala/dxWDL/base/Extras.scala:687]: That actually wasn't the IDE, that was me. I remember we had a conversation when I first started working on dxWDL when you asked me to always use curly braces even when they're not necessary. Maybe I over-interpreted that request.

File: src/main/scala/dxWDL/base/ParseWomSourceFile.scala [hunk 2]
```diff
@@ -155,31 +157,29 @@ case class ParseWomSourceFile(verbose: Boolean) {
...
-          System.out.println("invalid code generated ----- ")
-          System.out.println(wdlWfSource)
-          throw e
-      }
-    val typeInfer = TypeInfer(opts)
-    val (tDoc, _) = typeInfer.apply(doc)
+    val doc = parser.parseDocument(sourceCode)
+    val (tDoc, ctx) = TypeInfer(opts).apply(doc)
+    (tDoc, ctx)
+  }
```
[the remark this answers is not in the document]
Comment (jdidion) [reply src/main/scala/dxWDL/base/ParseWomSourceFile.scala:170]: I added the print statements back in.

**Comment by jdidion** [comment]
The changes are in many places, but mostly
* ParseWomSourceFile
* WdlCodeGen
* Generate*
* Native

File: src/main/scala/dxWDL/compiler/WdlCodeGen.scala [hunk 3]
```diff
@@ -45,234 +54,226 @@ case class WdlCodeGen(verbose: Verbose,
...
+        TAT.ExprArray(Vector.empty, wdlType, null)
 
       // Non empty array
       case WdlTypes.T_Array(t, true) =>
-        WdlValues.V_Array(Vector(genDefaultValueOfType(t)))
+        TAT.ExprArray(Vector(genDefaultValueOfType(t)), wdlType, null)
 
       case WdlTypes.T_Pair(lType, rType) =>
-        WdlValues.V_Pair(genDefaultValueOfType(lType), genDefaultValueOfType(rType))
+        TAT.ExprPair(genDefaultValueOfType(lType), genDefaultValueOfType(rType), wdlType, null)
```
Comment (orodeh) [inline src/main/scala/dxWDL/compiler/WdlCodeGen.scala:68]: I can see why you are using `null` for the TextSource. It is, generally not a good idea to use `null` in scala. Of course, this use may be justified. It is up to the programmer discretion. 

In similar cases, I created a dummy token with zeros for the line and column numbers.

File: src/main/scala/dxWDL/exec/TaskRunner.scala [hunk 4]
```diff
@@ -17,13 +17,14 @@ task Add {
...
+import java.nio.file.{Files, Path, Paths}
+
 import com.dnanexus.DXAPI
 import com.fasterxml.jackson.databind.JsonNode
 import java.lang.management._
-import java.nio.file.{Files, Path, Paths}
 import spray.json._
-import wdlTools.eval.{Context => EvalContext, WdlValues}
-import wdlTools.types.{TypedAbstractSyntax => TAT, WdlTypes}
+import wdlTools.eval.{WdlValues, Context => EvalContext}
```
Comment (orodeh) [inline src/main/scala/dxWDL/exec/TaskRunner.scala:26]: The idea is to sort the names alphabetically, so `WdlValues` comes after `Context`.

File: src/main/scala/dxWDL/compiler/GenerateIRTask.scala [hunk 5]
```diff
@@ -367,19 +364,34 @@ case class GenerateIRTask(verbose: Verbose,
...
+      case IR.DockerImageDxFile(_, dxFile) =>
         // The docker container is on the platform, we need to remove
         // the dxURLs in the runtime section, to avoid a runtime
         // lookup. For example:
         //
         //   dx://dxWDL_playground:/glnexus_internal  ->   dx://project-xxxx:record-yyyy
         val dxURL = DxUtils.dxDataObjectToURL(dxFile)
-        taskSourceCode.replaceAll(orgURL, dxURL)
-      case _ => taskSourceCode
+        task.copy(runtime = task.runtime.map(rt => replaceContainer(rt, dxURL)))
```
Comment (orodeh) [inline src/main/scala/dxWDL/compiler/GenerateIRTask.scala:389]: Thanks, this is much better, and much safer.

File: src/main/scala/dxWDL/compiler/Native.scala [hunk 6]
```diff
@@ -884,7 +880,7 @@ case class Native(dxWDLrtId: Option[String],
     val hintRunspec: Map[String, JsValue] = applet.runtimeHints match {
       case Some(hints) =>
         hints
-          .map {
+          .flatMap {
```
Comment (orodeh) [inline src/main/scala/dxWDL/compiler/Native.scala:883]: This IDE is going to kill me one day.

File: src/main/scala/dxWDL/exec/TaskRunner.scala (hunk 4 above)
Comment (jdidion) [reply src/main/scala/dxWDL/exec/TaskRunner.scala]: Fixed

File: src/main/scala/dxWDL/compiler/WdlCodeGen.scala (hunk 3 above)
Comment (jdidion) [reply src/main/scala/dxWDL/compiler/WdlCodeGen.scala]: yes, it was a bit lazy, but the code generator does not use the TextSource at all so in this case it is safe. I will make a note to go back and put in dummy values later.

**Comment by jdidion** [comment]
I have merged in your changes and all unit tests pass, so I will merge with TNG_II.

**Pull request merged**